pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/AvoidThrowingNullPointerExceptionRule.java

```diff
@@ -0,0 +1,79 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.rule.design;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import net.sourceforge.pmd.lang.java.ast.ASTAllocationExpression;
+import net.sourceforge.pmd.lang.java.ast.ASTClassOrInterfaceBody;
+import net.sourceforge.pmd.lang.java.ast.ASTClassOrInterfaceType;
+import net.sourceforge.pmd.lang.java.ast.ASTThrowStatement;
+import net.sourceforge.pmd.lang.java.ast.ASTVariableDeclarator;
+import net.sourceforge.pmd.lang.java.rule.AbstractJavaRule;
+
+/**
+ * Finds <code>throw</code> statements containing <code>NullPointerException</code>
+ * instances as thrown values
+ *
+ * @author <a href="mailto:michaeller.2012@gmail.com">Mykhailo Palahuta</a>
+ */
+public class AvoidThrowingNullPointerExceptionRule extends AbstractJavaRule {
+
+    @Override
+    public Object visit(ASTClassOrInterfaceBody body, Object data) {
+        List<ASTThrowStatement> throwNPEs = getThrowNullPointerExceptionStatements(body);
+        for (ASTThrowStatement throwNPE : throwNPEs) {
+            addViolation(data, throwNPE);
+        }
+        return data;
+    }
+
+    private List<ASTThrowStatement> getThrowNullPointerExceptionStatements(ASTClassOrInterfaceBody body) {
+        List<ASTThrowStatement> throwStatements = body.findDescendantsOfType(ASTThrowStatement.class);
+        List<String> npeInstances = getNullPointerExceptionInstances(body);
+        List<ASTThrowStatement> throwNPEStatements = new ArrayList<>();
+        for (ASTThrowStatement throwStatement : throwStatements) {
+            if (throwsNullPointerException(throwStatement, npeInstances)) {
+                throwNPEStatements.add(throwStatement);
+            }
+        }
+        return throwNPEStatements;
+    }
+
+    private List<String> getNullPointerExceptionInstances(ASTClassOrInterfaceBody body) {
+        List<ASTAllocationExpression> allocations = body.findDescendantsOfType(ASTAllocationExpression.class);
```

Comment:
We are finding here allocation in all methods. It could be a case, where one method creates a NPE and in another method some variable is thrown - it happens to have the same name (valid, because in a different method), but it might be a different exception, e.g.:

```java
public void bar() {
   NullPointerException e = new NullPointerException();
   e.printStackTrace();
}

public void foo() {
   RuntimeException e = new RuntimeException();
   throw e;
}
```

Could you please add this as an additional test case?

Maybe it's better to visit `ASTMethodDeclaration` and `ASTConstructorDeclaration` instead of the whole class via ASTClassOrInterfaceBody. If we change that, we need to add both classes as rule chain visits.